### DISCUSSION TITLE
question about the algorithm

### DISCUSSION THREAD

**Discussion by Vive940219** [voxelmorph/voxelmorph#242] [Q&A] [answered]
**Task** (what are you trying to do/register?)

I trying to register two knee images.

tried lddmm 

Please describe specifics of your approach // use of `vxm`

I have a question about the method its self. Is voxelmorph a elastic or fluid model? compared with lddmm
I think it combine the parameter based registration and fluid based registration. But I am not sure about this, because in normal voxelmorph without probabilistic diffeomoprhic , the loss function has no regularization of displacement or velocity. The loss fuction in model with probabilistic diffeomophic has the regularization of velocity. I am not totally understand the generative model in miccia 2018, but am I right?

**Comment by adalca** [comment, marked as answer]
@Vive940219 I think these terms are sort of abused in literature so I would avoid categorizing voxelmorph as one of them, especially since we tried to design voxelmorph as a generally useful registration library, but here's what i'd say

in the [TMI2019 paper](https://arxiv.org/abs/1809.05231), we **do** have a regularization, over the displacement field (loss = image_loss + \lambda * displacement_gradient). In the [probabilistic diffeomorphic model](https://arxiv.org/abs/1903.03545), the main regularization is on the velocity field. In our implementation, you can mix and match these easily -- for example, you the default now is to avoid the probabilistic angle (since most users don't care about getting uncertainties out), but it uses a velocity field. 

Does this answer your question?

**Comment by Vive940219** [comment]
@adalca Thanks for your answer. I think I missed the loss of gradient part in normal voxelmorph. I have another trouble about the probabilistic diffeomorphic model, compare with the normal voxelmorph, it adds a generative layer instead of the random initialization of the flow. I am  interested in this layer, can you give me more background behind this? In the miccai 2018, only 2.1 chapter mentioned this, I don't understand about why you introduce graph degree matrix and laplacian of neighborhood here. In your code it seems like you set  multiple gaussian kernels to estimate the initial of the flow. Excuse my troubling you, would your team upload the Pytorch vision of miccia 2018 recently?

**Comment by adalca** [comment]
@Vive940219 the generative modelling in the MICCAI paper (and maybe better explained with a figure in the _Medica Image Analysis_ paper here: https://arxiv.org/abs/1903.03545) allows us to describe relationships between variables (velocity field SVF, image noise, etc) with clean probablistic distribution, which is a nice way to reason about relationships of things you care about in the problem.

In particular, we can think of "prior knowledge" over velocity fields -- what do we know beforehand? one easy thing to know is that the field is expected to be 'smooth' in some sense, and one way to define smoothness is through a laplacian (which basically says that the SVF difference between neighbours is small). There's other ways to do this, but this was one (elegant) way. 

In the end, when deriving the loss, the KL (or alternatively you could derive a similar loss via the negative posterior), what this prior translates into is a term very similar to the spatial gradient of the velocity field. Since this is all it really boils down to, we have this implemented in pytorch -- since we have the `vxm.torch.losses.Grad` loss. We don't have the probabilistic aspect implemented in pytorch since it's not a priority right now (not our most requested feature) and we're just overwhelmed with feature requests and development of new projects like [SynthMorph](https://arxiv.org/pdf/2004.10282.pdf)

**Comment by Vive940219** [comment]
@adalca Thanks for your answer, in miccai2018 model, the velocity field is a stationary one, same exp integrator with vSVF. Compare with the time-varying velocity field e.g., LDDMM, it seems like the time-varying model can handle larger deformation(I am not sure here). In case that you mentioned that the volexmorph is a general registration method, can volexmorph performance better than vSVF,since we can control the size of kernels to handle larger deformation?
I am also curious about the displacement = exp(velocity),since we want a smooth field, you add a generative layer , in your code you also estimate the bidirection flow to add the loss term to ensure that the network can handle the inverse flow also. But this way a bidirection displacements needed. My question is that can we skip this estimation twice instead of using displacement = exp(-velocity) to generate the backward displacement field?

**Comment by adalca** [comment]
> @adalca Thanks for your answer, in miccai2018 model, the velocity field is a stationary one, same exp integrator with vSVF. Compare with the time-varying velocity field e.g., LDDMM, it seems like the time-varying model can handle larger deformation(I am not sure here). In case that you mentioned that the volexmorph is a general registration method, can volexmorph performance better than vSVF,since we can control the size of kernels to handle larger deformation?

I'm not sure that *in practice* the observation about time-varying VFs is true. It's certainly beautiful theory, but I'm not aware (and smarter people and I know better) of any work that shows this on imaging datasets.  I think SVfs (and voxelmorph using SVFs) can perform as well.

> I am also curious about the displacement = exp(velocity),since we want a smooth field, you add a generative layer , in your code you also estimate the bidirection flow to add the loss term to ensure that the network can handle the inverse flow also. But this way a bidirection displacements needed. My question is that can we skip this estimation twice instead of using displacement = exp(-velocity) to generate the backward displacement field?

Yes, I think you can. It might help a tiny bit because of interpolation effects (which are not taken into account in the theory), but it general we've found that the bidirectional loss has little effect for a large enough network trained to convergence.

p.s. closing this issue since i think the original question is solved, but feel free to continue the discussion here if you have more questions

**Comment by Vive940219** [comment]
@adalca Sorry that I have the last question https://arxiv.org/abs/1802.02604 in the chapter 4.1 you mentioned that "The receptive fields of the convolutional kernels of the smallest layer should be at least as larger as the maximum expectes diasplacement between corresponding voxels in M and F". why this limit? Because the largest displacement between M and F should be also downsampled at the smallest layer. Is the kernel size should at least equal to the MaxDis multiple by downsampling rate?

**Comment by adalca** [comment]
@Vive940219 yes, something like that. Basically the "largest displacement" should be within 1 or 2 voxels at the smallest layer (so maybe 16 or 32 or so at the highest level depending on your number of levels for usual brain mr data). So if you expect larger displacement, add more downsampling layer. Just as a guideline.

**Comment by voxelmorph** [comment]
@Vive940219 since GitHub enabled 'Discussions' and this was really a discussion rather than a code issue, I'm moving this thread to Discussions, FYI
